internal/addoninstancesdk/sdk.go

```diff
@@ -0,0 +1,168 @@
+	if addonInstanceHeartbeatReporterSingleton == nil {
+		return nil, fmt.Errorf("heartbeat-reporter not found to be initialised. Initialize it by calling `InitializeAddonInstanceHeartbeatReporterSingleton(...)`")
+	}
+	return addonInstanceHeartbeatReporterSingleton, nil
+}
+
+func (adihrClient AddonInstanceHeartbeatReporter) LatestCondition() metav1.Condition {
+	return adihrClient.latestCondition
+}
+
+func (adihrClient *AddonInstanceHeartbeatReporter) changeRunningState(desiredState bool) {
+	adihrClient.isRunningMutex.Lock()
+	adihrClient.isRunning = desiredState
+	adihrClient.isRunningMutex.Unlock()
+}
+
+func (adihrClient *AddonInstanceHeartbeatReporter) Start(ctx context.Context) error {
+	defer func() {
+		adihrClient.changeRunningState(false)
+	}()
+
+	// not allow the client/tenant to start multiple heartbeat reporter concurrently and cause data races
+	adihrClient.isRunningMutex.Lock()
+	if adihrClient.isRunning {
+		adihrClient.isRunningMutex.Unlock()
+		return fmt.Errorf("already found the heartbeat reporter to be running")
+	}
+	adihrClient.isRunning = true
+	adihrClient.isRunningMutex.Unlock()
+
+	adihrClient.ticker = time.NewTicker(adihrClient.currentInterval)
+	defer adihrClient.ticker.Stop()
+
+	for {
+		select {
+		case update := <-adihrClient.updateCh:
+			if update.newLatestCondition != nil {
+				// immediately register a new heartbeat upon receive one from the client/tenant side
+				if err := adihrClient.updateAddonInstanceStatus(ctx, *update.newLatestCondition); err != nil {
+					return fmt.Errorf("failed to update the addoninstance status: %w", err)
+				}
+				adihrClient.latestCondition = *update.newLatestCondition
+			}
+
+			// update the interval if the newInterval in the `update` is provided and is not equal to the existing interval
+			// synchronize the timer with this new interval
+			if update.newAddonInstanceSpec != nil {
+				if update.newAddonInstanceSpec.HeartbeatUpdatePeriod.Duration != adihrClient.currentInterval {
+					adihrClient.currentInterval = update.newAddonInstanceSpec.HeartbeatUpdatePeriod.Duration
+					adihrClient.ticker.Reset(adihrClient.currentInterval)
+				}
+			}
+		case <-adihrClient.ticker.C:
+			if err := adihrClient.updateAddonInstanceStatus(ctx, adihrClient.latestCondition); err != nil {
+				return fmt.Errorf("failed to report the regular heartbeat: %w", err)
+			}
+		case <-ctx.Done():
+			return fmt.Errorf("provided context exhausted")
+		case <-adihrClient.stopperCh:
+			return nil
+		}
+	}
+}
+
+func (adihrClient *AddonInstanceHeartbeatReporter) Stop() error {
+	select {
+	case adihrClient.stopperCh <- true:
+		return nil
+	case <-time.After(30 * time.Second):
+		return fmt.Errorf("failed to stop the reporter: timed out waiting for the reporter to ack the stop signal")
+	}
+}
+
+func (adihrClient *AddonInstanceHeartbeatReporter) SendHeartbeat(ctx context.Context, condition metav1.Condition) error {
```

Comment:
Good point. I would prefer it to be in this order:
```
SendHeartbeat(ctx, newConditions []metav1.Condition)
1. Perform the status update API call
2. updateCh <- newConditions
```

Though, what if the step 2 fails, then we should have some sort of rollback for Step 1. too. If we don't implement the rollback then, in the worst case scenario, it would take `StatusReporter.interval` seconds for the StatusReporter to synchronize the AddonInstance CR's Status.Conditions (in the cluster) to `StatusReporter.LatestConditions`, hence, unnecessarily causing eventual consistency.

It won't be a 100% full-proof mechanism coz in theory even the rollback could fail but that's still better than no rollbacks.